### PR TITLE
[ENH] Add PC algorithm

### PR DESCRIPTION
**Pull request by adam2392**
Fixes #11
Closes: #10 

Changes proposed in this pull request:

- PC algorithm implementation with separate skeleton discovery stage (has all the features of causal-learn I think(?), except for missing-value support)
- Adds the abstract class for Constraint Based learning algorithms
- Adds the `Context` class
- Adds an example showing how to run PC algorithm
- Adds a metric submodule for comparing graphs

## Before submitting

- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/dodiscover/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/dodiscover/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

**Comment by adam2392** [comment]
https://output.circle-artifacts.com/output/job/42854568-1d11-48d3-b89a-660449f4b9c0/artifacts/0/dev/index.html

**Comment by adam2392** [comment]
https://output.circle-artifacts.com/output/job/4ffd2aba-0ca5-4a98-ab4d-29c694a00382/artifacts/0/dev/index.html

**Comment by bloebp** [comment]
This PR is quite big. Would it be possible to have each of the listed changes as a separate PR? Otherwise it is almost impossible that there is a proper review. Optimally, a PR should not have more than ~250 lines (preferably less). Except this is simply moving stuff over that has been review before already.

**Comment by adam2392** [comment]
That's doable! The challenge here is that the CI is still evolving as well. Sorry some files are changing there. 

If you're okay with the skeleton, abstract class, etc. being PRed without tests first that's fine. 

Then the final PR for the PC algo would have the corresponding tests?

**Comment by bloebp** [comment]
You can also choose another PR (i.e. the branch) as a reference ('based on'), then if you make changes in, e.g., the CI stuff, it would be reflected here as well. Also, if there are inconsistencies, you can simply resolve them once the PR is approved. E.g., if the PR is about the PC-algorithm, we don't care about the details of the CI tests as long as we assume that there is 'something'. Then once the PR for the PC-algorithm is approved, you can simply adjust/base it to the latest changes of the CI tests.

Wouldn't it be possible to add tests for the classes independently of the PC algo? Just some basic tests (if they are necessary at all), like adding/deleting edges behave as expected etc. And then you can add further ones with the PC commit.

File: doc/api.rst [hunk 1]
```diff
@@ -34,4 +37,30 @@ Graph Protocols
    :toctree: generated/
 
    GraphProtocol
-   EquivalenceClassProtocol
\ No newline at end of file
+   EquivalenceClassProtocol
+   
+Constraint-Based Structure Learning
```
Comment (robertness) [inline doc/api.rst:42]: Shouldn't these have gone in the previous PR?  Perhaps they can go into a separate PR?

File: doc/conf.py [hunk 2]
```diff
@@ -242,16 +258,13 @@
     "image_scrapers": scrapers,
 }
 
-# Custom sidebar templates, maps document names to template names.
-html_sidebars = {
-    "index": ["search-field.html"],
-}
+# Add pygraphviz png scraper, if available
+try:
+    from pygraphviz.scraper import PNGScraper
```
Comment (robertness) [inline doc/conf.py:263]: Could graphviz visualiation be its own PR?

File: doc/whats_new/v0.1.rst [hunk 3]
```diff
@@ -28,6 +28,7 @@ Changelog
 
 - |Feature| Implement continuous integration and repository docs, by `Adam Li`_ (:pr:`15`)
 - |Feature| Implement conditional independence tests under the ``dodiscover.ci`` submodule, by `Adam Li`_ (:pr:`16`)
+- |Feature| Implement PC algorithm and skeleton for constraint-based structure learning algorithms under the ``dodiscover.constraint`` submodule, by `Adam Li`_ (:pr:`18`)
```
Comment (robertness) [inline doc/whats_new/v0.1.rst:31]: @adam2392 are you updating this change log manually?  Isn't this something we could handle with semantic commits?  @darthtrevino

Comment (adam2392) [reply doc/whats_new/v0.1.rst:31]: Yeah rn it's just a per PR changelog. What's semantic commits? :0

**Comment by adam2392** [comment]
> This PR is quite big. Would it be possible to have each of the listed changes as a separate PR? Otherwise it is almost impossible that there is a proper review. Optimally, a PR should not have more than ~250 lines (preferably less). Except this is simply moving stuff over that has been review before already.

@bloebp @robertness and @darthtrevino I've now broken up the PR a bit. I'm leaving this open to demonstrate the PC example for discussion, but for the code itself:

- first it would be good to merge #19 to consolidate issues with the CI
- second, I have added the skeleton learning method, which can be its own PR in #20 
- Once those are merged in, I can then add the abstract base class for constraint-based structure learning algos. #21
- Finally, the PC algo along with its tests and examples can be added in this PR after it is rebased on top of the `main` branch

**Comment by adam2392** [comment]
^ Apologies for the gigantic PRs. I am more used to doing things in bite-sized chunks, but getting a repo setup sometimes takes a bit of wrangling. Will be better about it :p 

Hopefully the breakups now make reviewing more streamlined and easy.

File: .circleci/config.yml [hunk 4]
```diff
@@ -48,28 +48,33 @@ jobs:
...
+          command: |
+            sudo apt install libspatialindex-dev xdg-utils
+      - python/install-packages:
+          pkg-manager: poetry
+          args: "--extras graph_func"
+          cache-version: "v1" # change to clear cache
+      - run:
+          name: Check poetry package versions
+          command: |
+            poetry run pip install --upgrade --progress-bar off pygraphviz
```
Comment (darthtrevino) [inline .circleci/config.yml:75]: This seems weird - why not add `pygraphviz` as a listed dependency in `pyproject.toml`? In general, my approach in dowhy has been to avoid `pip` entirely and model all dependencies inside of `pyproject.toml'. When we have these tacit installs it makes setting up a dev environment kind of tricky.

File: doc/whats_new/v0.1.rst (hunk 3 above)
Comment (darthtrevino) [reply doc/whats_new/v0.1.rst:31]: https://python-semantic-release.readthedocs.io/ - this library will parse git commit messages for changelog information

https://pypi.org/project/semversioner/ - this project will let you batch up change documents that are applied into the changelog via their CLI.

I kind of prefer the latter approach; commit-message parsing can be flaky if contributors aren't aware of the requirement. We use similar tooling in our JS projects, and we require that every PR has a semver-impact document associated with it.

**Review by darthtrevino — APPROVED** [review]
LGTM; just a question about pygraphviz install

File: .circleci/config.yml (hunk 4 above)
Comment (adam2392) [reply .circleci/config.yml:75]: Good point. I added it to the `pyproject.toml` and updated the poetry.lock file in #19

File: doc/whats_new/v0.1.rst (hunk 3 above)
Comment (robertness) [reply doc/whats_new/v0.1.rst:31]: > commit-message parsing can be flaky

I've certainly been guilty of crummy semantic commit messages.  I'm inclined to go with whatever you think is best @darthtrevino .  I'd like to employ what CI tech that can help these PR's can get leaner.  The leaner they are, the easier it is for me to onboard new contributors.

**Review by robertness — APPROVED** [review]

**Comment by adam2392** [comment]
Closed in favor of #30
